src/common/transformations/src/transformations/op_conversions/group_query_attention_decomposition.cpp

```diff
@@ -69,51 +73,60 @@ ov::OutputVector ov::pass::GroupQueryAttentionDecomposition::decompose(
+    const auto two = register_new_node(v0::Constant::create(ov::element::i64, ov::Shape{1}, {2}));
+    const auto seqlens_elemi64 = register_new_node<v0::Convert>(seqlens_k, ov::element::i64);
+    const auto real_seqlens = register_new_node<v1::Add>(seqlens_elemi64, one);
 
     // Only consider batch is 1
-    auto seqlens_1d = register_new_node<v1::Reshape>(real_seqlens, one, false);
-    auto past_sequence_length = register_new_node<v1::Subtract>(seqlens_1d, current_sequence_length);
+    const auto seqlens_1d = register_new_node<v1::Reshape>(real_seqlens, one, false);
+    const auto past_seqlen = register_new_node<v1::Subtract>(seqlens_1d, current_seqlen);
+    const auto curr_seqlen_scalar = register_new_node<v0::Squeeze>(current_seqlen);
+
     if (do_rotary) {
-        Q = rotaryEmbedding(Q,
-                            past_sequence_length,
-                            seqlens_1d,
-                            cos_cache.get_node_shared_ptr(),
-                            sin_cache.get_node_shared_ptr(),
-                            head_size_node,
-                            rotary_interleaved);
-        K = rotaryEmbedding(K,
-                            past_sequence_length,
-                            seqlens_1d,
-                            cos_cache.get_node_shared_ptr(),
-                            sin_cache.get_node_shared_ptr(),
-                            head_size_node,
-                            rotary_interleaved);
+        ov::Output<ov::Node> position_ids =
+            register_new_node<v4::Range>(zero_without_shape, curr_seqlen_scalar, one_without_shape, ov::element::i64);
+        position_ids = register_new_node<v1::Add>(position_ids, past_seqlen);
+
+        const auto cos = register_new_node<v8::Gather>(cos_cache, position_ids, zero);
+        const auto sin = register_new_node<v8::Gather>(sin_cache, position_ids, zero);
+        Q = rotaryEmbedding(Q, cos, sin, rotary_interleaved);
+        K = rotaryEmbedding(K, cos, sin, rotary_interleaved);
     }
 
     auto construct_kv_cache = [&](const ov::Output<ov::Node>& past, const ov::Output<ov::Node>& current) {
-        auto past_datas = register_new_node<v8::Slice>(past, zero, past_sequence_length, one, two);
-        auto curr_datas = register_new_node<v8::Slice>(current, zero, current_sequence_length, one, two);
-        return register_new_node<v0::Concat>(ov::NodeVector{past_datas, curr_datas}, 2);
+        return register_new_node<v0::Concat>(ov::OutputVector{past, current}, 2);
     };
     K = construct_kv_cache(past_key, K);
     V = construct_kv_cache(past_value, V);
-    auto present_k = K;
-    auto present_v = V;
 
+    const auto concat_kv_len = get_dimensions(K.get_node_shared_ptr(), {2});
+    const auto concat_kv_len_scalar = register_new_node<v0::Squeeze>(concat_kv_len);
+
+    ov::Output<ov::Node> present_k;
+    ov::Output<ov::Node> present_v;
+    if (is_static_input) {
```

Comment:
can we use "ShapeOf -> Gather" to cut K.get_partial_shape()[2].get_length() and cover dynamic case?
for static case "ShapeOf -> Gather" will be const folded, for dynamic case "ShapeOf -> Gather" remains in the graph

Reply:
Such case seems to be already handled by `get_dimensions` helper and stored as `concat_kv_len`, 
Looks like the idea of this change, is to introduce different behavior for static and dynamic shapes intentionally, and treat static shape as indicator of "maximum sequence length", while dynamic as the actual size that is changing between inference calls.

Reply:
The NPU doesn't support dynamic shape, which causes a runtime issue when doing type inference. At the beginning of the implementation, we used gatherOp, but the SliceOp can get better performance.